packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+  const base: any = isOptional ? (validator as any).value : validator;
+  let zodValidator: z.ZodTypeAny;
+  switch (base.kind) {
+    case "id":
+      zodValidator = zid((base as any).tableName);
+      break;
+    case "string":
+      zodValidator = z.string();
+      break;
+    case "float64":
+      zodValidator = z.number();
+      break;
+    case "int64":
+      zodValidator = z.bigint();
+      break;
+    case "boolean":
+      zodValidator = z.boolean();
+      break;
+    case "null":
+      zodValidator = z.null();
+      break;
+    case "any":
+      zodValidator = z.any();
+      break;
+    case "array":
+      zodValidator = z.array(convexToZod((base as any).element));
+      break;
+    case "object": {
+      const fields = (base as any).fields as Record<string, GenericValidator>;
+      const out: Record<string, z.ZodTypeAny> = {};
+      for (const [k, v_] of Object.entries(fields)) out[k] = convexToZod(v_);
+      zodValidator = z.object(out);
+      break;
+    }
+    case "union": {
+      const members = (base as any).members as GenericValidator[];
+      const zs = members.map((m) => convexToZod(m));
+      if (zs.length === 0) {
+        zodValidator = z.any();
+      } else if (zs.length === 1) {
+        zodValidator = zs[0] as z.ZodTypeAny;
+      } else {
+        const first = zs[0]!;
+        const second = zs[1]!;
+        const rest = zs.slice(2) as [z.ZodTypeAny?, ...z.ZodTypeAny[]];
+        zodValidator = z.union([first, second, ...rest.filter(Boolean) as z.ZodTypeAny[]]);
+      }
+      break;
+    }
+    case "literal":
+      zodValidator = z.literal((base as any).value);
+      break;
+    case "record":
+      // Restrict keys to string schema for compatibility
+      zodValidator = z.record(z.string(), convexToZod((base as any).value));
+      break;
+    default:
+      throw new Error(`Unknown convex validator: ${base.kind}`);
+  }
+  return isOptional ? z.optional(zodValidator) : zodValidator;
+}
+
+export function convexToZodFields(fields: PropertyValidators) {
+  const out: Record<string, z.ZodTypeAny> = {};
+  for (const [k, v_] of Object.entries(fields)) out[k] = convexToZod(v_ as GenericValidator);
+  return out as { [k in keyof typeof fields]: z.ZodTypeAny };
+}
+
+// Builders
+type OneArgArray<ArgsObject extends DefaultFunctionArgs = DefaultFunctionArgs> = [ArgsObject];
+type NullToUndefinedOrNull<T> = T extends null ? T | undefined | void : T;
+type Returns<T> = Promise<NullToUndefinedOrNull<T>> | NullToUndefinedOrNull<T>;
+
+type ReturnValueInput<ReturnsValidator extends z.ZodTypeAny | ZodValidator | void> = [
+  ReturnsValidator,
+] extends [z.ZodTypeAny]
+  ? Returns<z.input<ReturnsValidator>>
+  : [ReturnsValidator] extends [ZodValidator]
+  ? Returns<z.input<z.ZodObject<ReturnsValidator>>>
+  : any;
+
+// (unused) type kept for reference in v3 version
+// type ReturnValueOutput<...> omitted in zod4 to reduce type depth
```

Comment:
As @natedunn indicated I think that this type is necessary if we want for the functions to have the right return value types without needing an explicit `returns` validator